pennylane/ops/functions/bind_new_parameters.py

```diff
@@ -0,0 +1,129 @@
+# limitations under the License.
+"""
+This module contains the qml.bind_new_parameters function.
+"""
+
+from typing import Optional, Sequence
+import copy
+from functools import singledispatch
+
+import pennylane as qml
+from pennylane.typing import TensorLike
+from pennylane.operation import Operator, Tensor
+
+from ..op_math import CompositeOp, SymbolicOp, ScalarSymbolicOp
+
+
+def _validate_params(op: Operator, params: Optional[Sequence[TensorLike]]) -> Sequence[TensorLike]:
+    """Validate that the list of new parameters has the correct length and shape.
+
+    Args:
+        op (.Operator): Operator to update
+        params (Optional[Sequence[TensorLike]]): New parameters to create operator with
+
+    Returns:
+        Sequence[TensorLike]: Validated parameters
+
+    Raises:
+        ValueError: If the new parameters don't have the expected shape.
+    """
+    if params is None:
+        params = []
+
+    if isinstance(op, qml.Hamiltonian):
+        invalid_params = len(params) != len(op.coeffs)
+    else:
+        invalid_params = len(params) != op.num_params
+
+    if invalid_params:
+        raise ValueError(
+            "The length of the new parameters does not match the expected shape; "
+            f"got {len(params)}, expected {op.num_params}."
+        )
+
+    return params
+
+
+@singledispatch
+def bind_new_parameters(op: Operator, params: Optional[Sequence[TensorLike]] = None) -> Operator:
+    """Create a new operator with updated parameters
+
+    This function takes an :class:`~.Operator` and new parameters as input and
+    returns a new operator of the same type with the new parameters. This function
+    does not mutate the original operator.
+
+    Args:
+        op (.Operator): Operator to update
+        params (Sequence[TensorLike]): New parameters to create operator with
+
+    Returns:
+        .Operator: New operator with updated parameters
+
+    Raises:
+        ValueError: If the shape of the old and new operator parameters don't match
+    """
+    params = _validate_params(op, params)
+
+    return op.__class__(*params, wires=op.wires, **copy.deepcopy(op.hyperparameters))
+
+
+# pylint: disable=missing-docstring
+@bind_new_parameters.register
+def bind_new_parameters_composite_op(op: CompositeOp, params: Optional[Sequence[TensorLike]] = None):
+    params = _validate_params(op, params)
+    new_operands = []
+
+    for operand in op.operands:
+        sub_params = params[: operand.num_params]
+        params = params[operand.num_params :]
+        new_operands.append(bind_new_parameters(operand, sub_params))
+
+    return op.__class__(*new_operands)
+
+
+# pylint: disable=missing-docstring
+@bind_new_parameters.register
+def bind_new_parameters_symbolic_op(op: SymbolicOp, params: Optional[Sequence[TensorLike]] = None):
+    params = _validate_params(op, params)
+    new_scalar = []
+
+    if isinstance(op, ScalarSymbolicOp):
+        new_scalar.append(params[0])
+        params = params[1:]
```

Comment:
Could just have a separate registered method for `ScalaraSymbolicOp`?

Reply:
I think it might be worth it. I'll add that change.

Reply:
I've tried out creating a new method for `ScalarSymbolicOp`, and it makes the code even more complicated because of two reasons:

* Some of the subclasses of `ScalarSymbolicOp` take their arguments in the order `base, scalar`, and others take it in `scalar, base`. This is why I have a try-except block for the return statement of this function.
* `Pow` doesn't treat its scalar as a parameter, so its "parameters" that `bind_new_parameters` changes are only the parameters of its base. Because of this exception, I have to handle `Pow` in a different way, and doing so causes the same trouble that `Adjoint` is causing with `__new__`.

For now, I'm going to leave the code as is and it can be changed later if needed.

Reply:
Custom methods for each of the options?

That way we won't have to copy the base if its a member of the `hyperparameters`. Actually, do we need to deepcopy the hyperparameters at all?

Reply:
For operations like `Controlled` we do, since that's where control values, work wires, etc. are stored. I think it's good practice to copy the hyperparameters in case any new operations that inherit from `SymbolicOp` are added.